### PR TITLE
Allow name-based extractors to be irrefutable

### LINKED ISSUE #12240: Vector.unapplySeq is refutable
## reproduction steps

using Scala 2.13.4

```scala
scala> Vector(1) match {
  case Vector(_ @ _*) => ???
}
```

## problem

> match may not be exhaustive.
> It would fail on the following input: `Vector1()`

The match is totally exhaustive. Unfortunately, scalac doesn't know. `Vector.unapplySeq` seems to be using name-based pattern matching to return a `SeqFactory.UnapplySeqWrapper`. But name-based pattern matching can't AFAIK provide irrefutable patterns, so the pattern ends up refutable, and the warning (with the somewhat odd counterexample) is emitted.

### LINKED ISSUE #12254: `true` is not an irrefutable extractor type
In fact, it's not an extractor type at all:

```scala
object Test {
  sealed trait Foo
  final class Bar extends Foo

  object Bar {
    def unapply(o: Bar) : true = true
  }

  def f(foo: Foo) = foo match {
    case Bar() => println("Bar")
  }
}

Test.scala:10: error: The result type of an unapply method must contain a member `get` to be used as an extractor pattern, no such member exists in true
    case Bar() => println("Bar")
         ^
1 error
```

_Originally posted by @robby-phd in https://github.com/scala/bug/issues/12232#issuecomment-735774442_

### PR DESCRIPTION
**Pull request by martijnhoekstra**
Extractors ([SLS 2.13 8.1.9](https://www.scala-lang.org/files/archive/spec/2.13/08-pattern-matching.html#extractor-patterns)) are objects with an `unapply` and/or `unapplySeq` methods that can be used as patterns in match statements.

Name-based extractors are extractors where the `unapply`  method doesn't return `Option` but some other type, with a contract that it has (1) a method `get` returning the extracted thing and (2) a method `isEmpty` returning `false` if the extractor matches and `true` if it doesn't.

Prior to this change, having the static return type `Some`  instead of `Option` for `unapply` indicated that the extractor is "irrefutable", meaning the compiler knows the pattern always matches, which it uses for reachability and exhaustiveness analysis. This PR adds two more forms of irrefutable extractors:

1. "Boolean Match" extractors, where `unapply` returns the singleton type `true` instead of `Boolean`
2. Name-based extractors, where `isEmpty` returns the singleton type `false` instead of `Boolean`

With these changes, using such extractors removes what would otherwise be spurious false positive exhaustivity warnings.

---

PR implementation notes:

Irrefutable name-based extractors on `isEmpty: false`

* [x] ~Check/update spec~ wontfix in this PR. Neither name-based extractors nor irrefutable extractors are specced.
* [x] Align with dotty
* [x] more tests, both neg and pos
* [x] check if this removes the necessity for List special casing
* [x] add tests from scala/bug#12240, which is the motivating bug

Dotty: https://github.com/lampepfl/dotty/blob/cda0a9d496cc8fd6c10cb4c90c0a521a2c6820df/compiler/src/dotty/tools/dotc/transform/patmat/Space.scala#L292-L320 : constant false isEmpty is irrefutable

List special casing: Will still be needed as is, probably won't be needed anymore if we can check if all arities of `unapplySeq` have cases.

spun out the added testcases in 12240 to their own bug in scala/bug#12252

Fixes scala/bug#12240
Fixes scala/bug#12254

### CODE REVIEW

File: src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala
Comment (dwijnand) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala]: ConstantFalse

**Comment by dwijnand** [comment]
TODO: add tests from https://github.com/scala/bug/issues/12240?

**Comment by martijnhoekstra** [comment]
Supporting exhaustiveness/unused checking over multiple arities of unapplySeq extractors is quite a bit more complicated: if element arity exhaustiveness is checked at all yet, I would appreciate a pointer to it. Otherwise, the ambition level of covering understanding that `case Vector()`, `case Vector(_)` and `case Vector(_, _, _*)` together cover all of it is probably too high.

File: src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala
Comment (martijnhoekstra) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala:380]: @dwijnand how should I do this to be not wrong? I tried the following, which trips the assertion:

```
      private def isIrrefutabilityProof(sym: Symbol): Boolean = {
        def isConstantFalse1 = sym.asMethod.returnType match {
          case c: ConstantType => c.value == Constant(false)
          case _ => false
        }
        def isConstantFalse2 = sym.asMethod.returnType == ConstantFalse
        sym.isMethod && sym.name == TermName("isEmpty") && {
          assert(isConstantFalse1 == isConstantFalse2, s"symbol $sym has inconsistent return type ${sym.asMethod.returnType}")
          isConstantFalse1
        }
      }
```

**Comment by martijnhoekstra** [comment]
now also fixes https://github.com/scala/bug/issues/12254:

allow `true` as an extractor type, and make it irrefutable

**Comment by dwijnand** [comment]
> [@martijnhoekstra](/martijnhoekstra) [Martijn Hoekstra](/martijnhoekstra) marked this pull request as ready for review [3 days ago](#event-4048035726)

Obviously still need to get CI passing, but more generally: let me know when this is ready to review (and merge) as I'm thinking it would be simpler to have this merged and fix the other patmat issues on top, rather than in parallel to your workings.

**Comment by martijnhoekstra** [comment]
Oh, I didn't see that the `: true` change broke a neg test output -- `throw new Exception` now infers as a valid extractor type as subtype of `Boolean`

I -update-check'ed that, and for pushed the fixup.

Happy for review in either state. Whether the error message should complain about an extractor of `Nothing` not having a `get` method as a name-based extractor or not supporting 2 extracted members as a Boolean extractor probably doesn't really matter in the end.

**Comment by martijnhoekstra** [comment]
Jenkins likes it, so have at it @dwijnand 

I'm sure Travis will concur any day now.

**Comment by SethTisue** [comment]
@martijnhoekstra mind editing the PR description to begin with a section of user-facing documentation of what changed here, to be linked from the release notes? the existing, more reviewer-facing text can remain at the bottom

File: test/files/neg/t12240.scala
Comment (SethTisue) [inline test/files/neg/t12240.scala]: typo :-)

File: test/files/neg/t12240.scala
Comment (SethTisue) [inline test/files/neg/t12240.scala]: typo :-)

File: test/files/pos/t12240.scala
Comment (dwijnand) [inline test/files/pos/t12240.scala]:
```suggestion
    case Vector(_*) => "scalac doesn't know the match is already exhaustive"
```

File: test/files/neg/t4425b.check
Comment (dwijnand) [inline test/files/neg/t4425b.check]: Definitely we should keep the previous warning.  I think you can use `isNonBottomSubClass` instead of `<:<`, but maybe you already had a plan.

**Review by dwijnand — CHANGES_REQUESTED** [review]
> Whether the error message should complain about an extractor of `Nothing` not having a `get` method as a name-based extractor or not supporting 2 extracted members as a Boolean extractor probably doesn't really matter in the end.

Probably not, but looks like a bad change.

File: test/files/neg/t4425b.check
Comment (martijnhoekstra) [reply test/files/neg/t4425b.check]: Or should it maybe be note that it can't be `Nothing`?

Or should unapply methods returning `Nothing` maybe just work, and non-termninate at runtime? That's the most principled but also probably the least useful thing to do.

Comment (dwijnand) [reply test/files/neg/t4425b.check]: Hmm.  Maybe it should??  Say I temporarily redefine an extractor with `???` do I expect all my extractor usages to error out?  Probably not.  So maybe you're right that this should work...  But I'm not 100% sure.

Comment (martijnhoekstra) [reply test/files/neg/t4425b.check]: I'm leaning mostly towards `def unapply = ???` being it's own custom error, something along the lines of "extractors may not return Nothing from their unapply method".

That way, it leads you on the trail of what you should do (e.g. annotate as `Option[Extracted]` while keeping the body `???`).

All other options seem less helpful. I'm warming up to the concept of just allowing it as an extractor for any arity and type, but that's probably more work than the effort is worth.

To me, the message that Nothing doesn't have a get method is the yankiest, closely followed by the message this PR makes instead, but in the end it's a pretty exotic use case anyway and a bit of yank gives flavour.

Anyway, your call on what to go for within the scope of this PR.

Comment (dwijnand) [reply test/files/neg/t4425b.check]: I grokked what it was saying and why, but it wasn't obvious.  The "expected 0, found 2" is misleading IMO.  The reference to Nothing in the previous makes it a better state IMO.  But I do agree it's an exotic area.

Hmm, I'll let it stew for a couple of days.  See if anyone else wants to chime in.  Otherwise, I'm really happy to get these fixes in, thank you!!

Comment (martijnhoekstra) [reply test/files/neg/t4425b.check]: updated to "object X can't be used as an extractor: The result type of an unapply method may not be Nothing"

File: test/files/run/patmat-seq.check
Comment (dwijnand) [inline test/files/run/patmat-seq.check:2]: `A` and `B` are:

```scala
object A {
  def unapplySeq(a: Int) = new collection.SeqFactory.UnapplySeqWrapper(collection.mutable.ArrayBuffer(1,2,3))
}

object B {
  def unapplySeq(a: Int) = Some(collection.immutable.ArraySeq(1,2,3))
}
```

and the match is:

```scala
2 match {
  case A(xs @ _*)    => xs      // wrapper.get.toSeq
  case A(x, y)       => (x, y)  // wrapper.get.apply(0/1)
  case A(x, xs @ _*) => (x, xs) // wrapper.get.drop(1)
  case B(xs @ _*)    => xs
  case B(x, y)       => (x, y)
  ...
}
```

So previously it had to get to `B` (which returns `Some`) before figuring out that the second `B` pattern is unreachable.  Now it understands that the first unapplySeq covers everything and all the remaining patterns are unreachable. 👍🏼

File: test/files/run/patmat-seq.check
Comment (dwijnand) [inline test/files/run/patmat-seq.check]: What if it's `def isEmpty: false = { println("boo"); false }`?

File: test/files/run/patmatnew.check
Comment (dwijnand) [inline test/files/run/patmatnew.check:22]:
```scala
def doMatch(xs: List[String]): String = xs match {
  case List(_*) => "ok"
}
```

No longer considered inexhaustive 👍🏼

File: test/files/run/t3530.check
Comment (dwijnand) [inline test/files/run/t3530.check:4]:
```scala
def f2[T](x: List[T]) = println(x match {
  case List(_, _)       => "two"
  case List(_, _, _)    => "three"
  case List(xs @ _*)    => "list: " + xs.length
  // bug: the default case is marked unreachable
  // case _                => "not a list"
})
```

No longer considered inexhaustive 👍🏼

**Review by dwijnand — CHANGES_REQUESTED** [review]
I think this is good (I started on Thursday, took 3 days off and came back today) with exception to the `isEmpty` inlining, which is incorrect to do.

File: test/files/run/patmat-seq.check
Comment (dwijnand) [reply test/files/run/patmat-seq.check]: Btw, I fully expect this to be a pre-existing bug, basically unrelated to your PR here...

Comment (martijnhoekstra) [reply test/files/run/patmat-seq.check]: I'd have to check.

Comment (martijnhoekstra) [reply test/files/run/patmat-seq.check]: see the new tests: This doesn't seem to break anything, but I wouldn't know why.

Comment (dwijnand) [reply test/files/run/patmat-seq.check]: Hmm.  Maybe copy how `test/files/run/patmat-seq.scala` is setup to look at the trees.

Comment (martijnhoekstra) [reply test/files/run/patmat-seq.check]: I'll give that a try, I'll ping you when I have something like that.

Comment (martijnhoekstra) [reply test/files/run/patmat-seq.check]: I've added a new test along these lines, but I can't manage to actually get it to emit the same tree. Maybe it's special-cased for UnnaplySeqWrapper somewhere? If it is, I can't find where.

Comment (martijnhoekstra) [reply test/files/run/patmat-seq.check]: The new test "shows" the call doesn't get elided. Unfortunately, I can't confirm that it only inlines the check if it's safe, since I can't get it to inline at all.

Comment (dwijnand) [reply test/files/run/patmat-seq.check]: Are you working on this?  Should I jump in?  I'm not 100% sure when 3.0.0-M4 or 3.0.0-RC1 are coming out, but we want to release 2.13.5 near it and I'm keen for these improvements to make it in.

Comment (martijnhoekstra) [reply test/files/run/patmat-seq.check]: Can you take a look whether this is enough for you to satisfy the fear of a mis-inlined `isEmpty` side-effect? If not, do you have any idea how I can trigger that inlining seen in this testcase?

Comment (dwijnand) [reply test/files/run/patmat-seq.check]: Well that was painful...  and I'm not sure if the way I "fixed" it is reasonable/safe or not.

File: src/compiler/scala/tools/nsc/transform/patmat/MatchCodeGen.scala
Comment (dwijnand) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchCodeGen.scala]: @lrytz @retronym does this look reasonable to you?  The relevant adaptation line is https://github.com/scala/scala/blob/f7d0e1fd8768cfc0d7fa7237f900e68d6d31d92a/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L1249 and I was using [`882972d` (#7620)](https://github.com/scala/scala/pull/7620/commits/882972d3a850887210b5c9041ec9b4d9557cfc2e) as a reference, but I couldn't find a way to do something equivalent to the `deriveCaseDef`/`deconst` it does there, so I ended up with this.

File: src/compiler/scala/tools/nsc/transform/patmat/MatchCodeGen.scala
Comment (dwijnand) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchCodeGen.scala]: This is `isEmptySym`

**Comment by martijnhoekstra** [comment]
have you been able to find the conditions under which isEmpty is now constant-folded yet?

**Comment by dwijnand** [comment]
Oh yes.  It's constant folded because typer constant folds everything when past typer phase, and the pattern matcher has to get the trees it creates typed (as well as named, actually):

https://github.com/scala/scala/blob/bda011524f7619b61b4fcc8172f088d2445e947e/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L1249-L1250

Those conditions would be enough to get it to deconst there but, as I said, we're in patmat phase so way past typer.  I don't know enough about phase traveling to know if it would be sane or even work if we made the patmat tree typing pretend to happen during the real typer phase...  So I went with this instead.  Unfortunately I had to revert 2 pre-existing checkfiles that your patch had fixed, but at least your tests pass.  Just waiting for confirmation from the CIs.

**Comment by dwijnand** [comment]
I'm using https://github.com/scala/scala/commit/882972d3a850887210b5c9041ec9b4d9557cfc2e as guide and precedent for doing some pretty typer-y thing in patmat (`defaultLabel setInfo MethodType(Nil, if (b.tpe != null) b.tpe.deconst else pt)`)

**Comment by martijnhoekstra** [comment]
what's so weird is that obviously the isEmpty of UnapplySeqWrapper was constant folded, but I wasn't able to construct any other source where isEmpty was getting constant folded.

**Comment by dwijnand** [comment]
See my `run/patmat-no-inline-isEmpty.scala` test: it only occurs under separate compilation, that's why I compile twice.  I didn't get into why that was true or if it should be true.

**Comment by martijnhoekstra** [comment]
Ah, I see! That's the darndest thing -- and scary too. Maybe it doesn't just apply to this situation.

**Review by dwijnand — APPROVED** [review]

**Comment by dwijnand** [comment]
Lukas wondered if the unapply could be constant folded too, so I added a similar test for that and it's not constant folded - so all is good.

File: src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala]: `nme.isEmpty`?

File: src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala]: Maybe you were comparing the types with `==`? On types that's reference equality, for type equality you'd have to use `=:=`. Anyway, the current code LGTM, so I'd remove the comment, it's a bit confusing.

File: src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala]: Would / should `asMethod` work with if it's defined as `val isEmpty` (instead of `def`)? Maybe use `sym.tpe.finalResultType`?

File: src/compiler/scala/tools/nsc/transform/patmat/PatternExpansion.scala
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/patmat/PatternExpansion.scala]: Alternatively: `unapplyResultType().typeSymbol == definitions.BooleanClass`

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: "keep constant types when they are safe to fold. erasure eliminates constant types modulo some exceptions, so keep those."

**Review by lrytz — APPROVED** [review]
Just some minor suggestions.

File: src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala
Comment (dwijnand) [reply src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala]: I think this is related to the separate vs joint compilation issue of constant types.  ConstantFalse is the foldable kind and `neg/t12240.scala` defines `def isEmpty: false` which, IIUC, is a (non-foldable) literal type in joint compilation and only a foldable constant type if it were to be used in a separate compilation.

File: src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala
Comment (dwijnand) [reply src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala]: Looks like val defs are also method symbols?

```
scala> class Foo { val isEmpty: false = false }
class Foo

scala> :power
Power mode enabled. :phase is at typer.
import scala.tools.nsc._, intp.global._, definitions._
Try :help or completions for vals._ and power._

scala> symbolOf[Foo].tpe.member(TermName("isEmpty")).asMethod
val res0: $r.intp.global.MethodSymbol = value isEmpty
```

Comment (lrytz) [reply src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala]: That's what I expected (it uses the accessor). Add a test?

Comment (dwijnand) [reply src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala]: I changed it your way :)

**Comment by SethTisue** [comment]
🎉 👏

**Pull request merged**
